opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -191,33 +202,140 @@ private Optional<ExprType> type(String field) {
+   * @param dateType - field type
+   * @return Instant without timezone
    */
-  private ExprValue constructTimestamp(String value) {
-    try {
-      return new ExprTimestampValue(
-          // Using OpenSearch DateFormatters for now.
-          DateFormatters.from(DATE_TIME_FORMATTER.parse(value)).toInstant());
-    } catch (DateTimeParseException e) {
-      throw new IllegalStateException(
+  private ExprValue parseTimestampString(String value, OpenSearchDateType dateType) {
+    Instant parsed = null;
+    for (DateFormatter formatter : dateType.getAllNamedFormatters()) {
+      try {
+        TemporalAccessor accessor = formatter.parse(value);
+        ZonedDateTime zonedDateTime = DateFormatters.from(accessor);
+        // remove the Zone
+        parsed = zonedDateTime.withZoneSameLocal(ZoneId.of("Z")).toInstant();
+      } catch (IllegalArgumentException ignored) {
+        // nothing to do, try another format
+      }
+    }
+
+    // FOLLOW-UP PR: Check custom formatters too
+
+    // if no named formatters are available, use the default
+    if (dateType.getAllNamedFormatters().size() == 0
+        || dateType.getAllCustomFormatters().size() > 0) {
+      try {
+        parsed = DateFormatters.from(DATE_TIME_FORMATTER.parse(value)).toInstant();
+      } catch (DateTimeParseException e) {
+        // ignored
+      }
+    }
+
+    if (parsed == null) {
+      // otherwise, throw an error that no formatters worked
+      throw new IllegalArgumentException(
           String.format(
-              "Construct ExprTimestampValue from \"%s\" failed, unsupported date format.", value),
-          e);
+              "Construct ExprTimestampValue from \"%s\" failed, unsupported date format.", value)
+      );
+    }
+
+    return new ExprTimestampValue(parsed);
+  }
+
+  /**
+   * return the first matching formatter as a time without timezone.
+   *
+   * @param value - time as string
+   * @param dateType - field data type
+   * @return time without timezone
+   */
+  private ExprValue parseTimeString(String value, OpenSearchDateType dateType) {
+    for (DateFormatter formatter : dateType.getAllNamedFormatters()) {
+      try {
+        TemporalAccessor accessor = formatter.parse(value);
+        ZonedDateTime zonedDateTime = DateFormatters.from(accessor);
+        return new ExprTimeValue(
+            zonedDateTime.withZoneSameLocal(ZoneId.of("Z")).toLocalTime());
+      } catch (IllegalArgumentException  ignored) {
+        // nothing to do, try another format
+      }
+    }
+
+    // if no named formatters are available, use the default
+    if (dateType.getAllNamedFormatters().size() == 0) {
+      try {
+        return new ExprTimeValue(
+            DateFormatters.from(STRICT_HOUR_MINUTE_SECOND_FORMATTER.parse(value)).toLocalTime());
+      } catch (DateTimeParseException e) {
+        // ignored
+      }
     }
+    throw new IllegalArgumentException("Construct ExprTimeValue from \"" + value
```

Comment:
The same for exception - maybe return `NULL`?

Reply:
Or post format and value in the exception message

Reply:
Discussion to be had for a follow up PR including user defined formats